File: node/js/pure/identicon.js
// The root-derived identicon: the picture a persona wears when they haven't chosen one.
//
// Canon gives it a job (PROJECT_PLAN, Naming): "confusable-name attacks are further caught
// by the root-derived identicon - the name may collide; the image will not." That job only
// works if the image is the SAME everywhere the persona appears, so this is a twinned pure
// function (src/identicon.rs is the other half, held to the same goldens) rather than a
// library: no image crate can be reproduced bit-for-bit by a browser, and two faces that
// disagree defeat the whole point.
//
// No hash, deliberately. A root pubkey is already 32 uniformly-random bytes, so the picture
// reads the key's own bytes - which keeps this module dependency-free (the pure zone admits
// no imports) and makes the derivation obvious to anyone auditing it.
//
// The shape: a 5x5 grid, mirrored left-to-right (columns 0..2 are drawn, 3..4 reflect), so
// the figure is symmetric and reads as a face-ish glyph at 22px as well as 72px. Symmetry is
// what makes small identicons memorable, and mirroring means the hexagon's clipped corners
// take away the same thing on both sides.
import { personaHue } from './person.js';

const CELLS = 15; // 3 drawn columns x 5 rows; the other two columns mirror

/// The identicon as an SVG string, `viewBox`-scaled so one string serves every size.
/// Deterministic: same root, same bytes, same picture, in both languages.
export function identiconSvg(rootHex) {
    const hex = (rootHex || '').toLowerCase();
    const byte = (i) => parseInt(hex.slice(i * 2, i * 2 + 2), 16) || 0;
    const hue = personaHue(hex);
    // Three tones from the persona's own hue: the ring's colour is the ink, a shifted
    // sibling adds character, and a pale wash of the same family is the ground - so the
    // identicon and the hexagon's ring always read as one object.
    const ink = `hsl(${hue}, 62%, 42%)`;
    const accent = `hsl(${(hue + 42) % 360}, 68%, 58%)`;
    const ground = `hsl(${hue}, 34%, 92%)`;

    let cells = '';
    for (let i = 0; i < CELLS; i++) {
        const b = byte(i);
        // A DIFFERENT bit per cell, not always the low one: a key whose bytes share a
        // pattern (0xaa repeating, say) would otherwise draw the same answer fifteen times
        // and come out blank or solid. Half the cells fill, on average, either way.
        if (!((b >> (i % 7)) & 1)) continue;
        const col = Math.floor(i / 5);
        const row = i % 5;
        const fill = (b >> ((i + 3) % 7)) & 1 ? accent : ink; // another bit picks the tone
        cells += `<rect x="${col}" y="${row}" width="1" height="1" fill="${fill}"/>`;
        if (col < 2) {
            cells += `<rect x="${4 - col}" y="${row}" width="1" height="1" fill="${fill}"/>`;
        }
    }
    return (
        `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 5 5" shape-rendering="crispEdges">` +
        `<rect width="5" height="5" fill="${ground}"/>${cells}</svg>`
    );
}

/// The same picture as a data URI, for an `<img src>`. (The anonymous static face inlines
/// the SVG element instead - its CSP allows no data: images, and inlining needs no
/// permission at all.)
export function identiconUri(rootHex) {
    return `data:image/svg+xml,${encodeURIComponent(identiconSvg(rootHex))}`;
}

File: node/src/identicon.rs
//! The root-derived identicon, Rust half - twin of `js/pure/identicon.js`, held to the same
//! goldens. The anonymous face draws this for a persona with no picture of their own; the
//! console draws the identical SVG from the identical bytes, because an identicon that
//! differs between two views of one persona cannot do the job canon gives it (PROJECT_PLAN,
//! Naming: "the name may collide; the image will not").
//!
//! No hash and no image crate: a root pubkey is already 32 uniformly-random bytes, and a
//! crate's PNG could never be reproduced bit-for-bit in a browser.

/// The persona's hue - the same number `pure/person.js` derives, and the same one the
/// hexagon's ring wears.
pub fn hue(root: &[u8; 32]) -> u32 {
    (u32::from(root[0]) * 65536 + u32::from(root[1]) * 256 + u32::from(root[2])) % 360
}

/// The identicon as an SVG string: a 5x5 grid mirrored left-to-right, three tones from the
/// persona's own hue. `viewBox`-scaled, so one string serves every size.
pub fn identicon_svg(root: &[u8; 32]) -> String {
    let h = hue(root);
    let ink = format!("hsl({h}, 62%, 42%)");
    let accent = format!("hsl({}, 68%, 58%)", (h + 42) % 360);
    let ground = format!("hsl({h}, 34%, 92%)");

    let mut cells = String::new();
    // The first fifteen bytes are the fifteen drawn cells (3 columns x 5 rows); the other
    // two columns mirror them.
    for (i, &b) in root.iter().take(15).enumerate() {
        // A different bit per cell (see the JS twin): a patterned key must not draw the same
        // answer fifteen times.
        if (b >> (i % 7)) & 1 == 0 {
            continue;
        }
        let col = i / 5;
        let row = i % 5;
        let fill = if (b >> ((i + 3) % 7)) & 1 == 1 { &accent } else { &ink };
        cells.push_str(&format!(
            r#"<rect x="{col}" y="{row}" width="1" height="1" fill="{fill}"/>"#
        ));
        if col < 2 {
            cells.push_str(&format!(
                r#"<rect x="{}" y="{row}" width="1" height="1" fill="{fill}"/>"#,
                4 - col
            ));
        }
    }
    format!(
        r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 5 5" shape-rendering="crispEdges"><rect width="5" height="5" fill="{ground}"/>{cells}</svg>"#
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The cross-language goldens - identical strings pinned in
    /// integration/test/pure/identicon.cjs. Drift on either side fails before a persona
    /// shows two different faces.
    const GOLDENS: [(&str, &str); 2] = [
        (
            "93ad0ddd9dd2022bf2ac21664b386965e0eeffecaff6e49b71039db5f1cf53f3",
            r##"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 5 5" shape-rendering="crispEdges"><rect width="5" height="5" fill="hsl(213, 34%, 92%)"/><rect x="0" y="0" width="1" height="1" fill="hsl(213, 62%, 42%)"/><rect x="4" y="0" width="1" height="1" fill="hsl(213, 62%, 42%)"/><rect x="0" y="2" width="1" height="1" fill="hsl(213, 62%, 42%)"/><rect x="4" y="2" width="1" height="1" fill="hsl(213, 62%, 42%)"/><rect x="0" y="3" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="4" y="3" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="0" y="4" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="4" y="4" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="1" y="2" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="3" y="2" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="1" y="3" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="3" y="3" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="1" y="4" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="3" y="4" width="1" height="1" fill="hsl(255, 68%, 58%)"/><rect x="2" y="4" width="1" height="1" fill="hsl(255, 68%, 58%)"/></svg>"##,
        ),
        (
            "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            r##"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 5 5" shape-rendering="crispEdges"><rect width="5" height="5" fill="hsl(330, 34%, 92%)"/><rect x="0" y="1" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="4" y="1" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="0" y="3" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="4" y="3" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="1" y="0" width="1" height="1" fill="hsl(12, 68%, 58%)"/><rect x="3" y="0" width="1" height="1" fill="hsl(12, 68%, 58%)"/><rect x="1" y="3" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="3" y="3" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="2" y="0" width="1" height="1" fill="hsl(330, 62%, 42%)"/><rect x="2" y="2" width="1" height="1" fill="hsl(12, 68%, 58%)"/></svg>"##,
        )
    ];

    fn root(hex: &str) -> [u8; 32] {
        let mut out = [0u8; 32];
        for (i, byte) in out.iter_mut().enumerate() {
            *byte = u8::from_str_radix(&hex[i * 2..i * 2 + 2], 16).unwrap();
        }
        out
    }

    #[test]
    fn draws_the_goldens_exactly() {
        for (hex, svg) in GOLDENS {
            assert_eq!(identicon_svg(&root(hex)), svg, "identicon for {hex}");
        }
    }

    #[test]
    fn different_keys_draw_different_pictures() {
        assert_ne!(
            identicon_svg(&root(GOLDENS[0].0)),
            identicon_svg(&root(GOLDENS[1].0))
        );
    }
}
